plugins/intel-usb4/fu-intel-usb4-firmware.c

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (C) 2022 Richard Hughes <richard@hughsie.com>
+ * Copyright (C) 2021 Intel Corporation.
+ * Copyright (C) 2021 Dell Inc.
+ * All rights reserved.
+ *
+ * This software and associated documentation (if any) is furnished
+ * under a license and may only be used or copied in accordance
+ * with the terms of the license.
+ *
+ * This file is provided under a dual MIT/LGPLv2 license.  When using or
+ * redistributing this file, you may do so under either license.
+ * Dell Chooses the MIT license part of Dual MIT/LGPLv2 license agreement.
+ *
+ * SPDX-License-Identifier: LGPL-2.1+ OR MIT
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-intel-usb4-firmware.h"
+
+struct _FuIntelUsb4Firmware {
+	FuFirmwareClass parent_instance;
+	guint16 vendor_id;
+	guint16 product_id;
+};
+
+G_DEFINE_TYPE(FuIntelUsb4Firmware, fu_intel_usb4_firmware, FU_TYPE_FIRMWARE)
+
+/* NVM offset */
+#define NVM_VER_OFFSET_MAJOR 0xa
+#define NVM_VER_OFFSET_MINOR 0x9
+#define NVM_VID_OFFSET_MAJOR 0x221
+#define NVM_VID_OFFSET_MINOR 0x220
+#define NVM_PID_OFFSET_MAJOR 0x223
+#define NVM_PID_OFFSET_MINOR 0x222
+
+static gboolean
+fu_intel_usb4_firmware_parse(FuFirmware *firmware,
+			     GBytes *fw,
+			     gsize offset,
+			     FwupdInstallFlags flags,
+			     GError **error)
+{
+	FuIntelUsb4Firmware *self = FU_INTEL_USB4_FIRMWARE(firmware);
+	guint16 version_raw = 0x0;
+	guint32 hdr_offset = 0x0;
+	g_autofree gchar *version = NULL;
+	g_autoptr(FuFirmware) img_payload = NULL;
+	g_autoptr(GBytes) fw_payload = NULL;
+
+	/* get header offset */
+	if (!fu_memread_uint32_safe(g_bytes_get_data(fw, NULL),
+				    g_bytes_get_size(fw),
+				    0x0,
+				    &hdr_offset,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return FALSE;
+
+	/* vid:pid */
+	if (!fu_memread_uint16_safe(g_bytes_get_data(fw, NULL),
+				    g_bytes_get_size(fw),
+				    hdr_offset + NVM_VID_OFFSET_MINOR,
+				    &self->vendor_id,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return FALSE;
+	if (!fu_memread_uint16_safe(g_bytes_get_data(fw, NULL),
+				    g_bytes_get_size(fw),
+				    hdr_offset + NVM_PID_OFFSET_MINOR,
+				    &self->product_id,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return FALSE;
+
+	/* version */
+	if (!fu_memread_uint16_safe(g_bytes_get_data(fw, NULL),
+				    g_bytes_get_size(fw),
+				    hdr_offset + NVM_VER_OFFSET_MINOR,
+				    &version_raw,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return FALSE;
+	version = fu_version_from_uint16(version_raw, FWUPD_VERSION_FORMAT_PAIR);
```

Comment:
`version` here results in decimal from `%u`, maybe we want it in hex `%x`

Reply:
Ohh, indeed! Dell Dock uses %02x -- but is that a PAIR?

Reply:
yes, the format is a PAIR e.g. AA.BB

Reply:
Fixed, thanks.